File: implementations/elixir/lib/ockam/vault.ex
defmodule Ockam.Vault do
  @moduledoc """
  """

  alias Ockam.Vault.NIF

  defmodule Secret do
    @moduledoc false
    defstruct [:reference]

    @opaque t :: %__MODULE__{}
  end

  defmodule SecretAttributes do
    @moduledoc false
    defstruct ty: :buffer, length: 0, persistence: :ephemeral, purpose: :key_agreement
  end

  defstruct [:reference]

  @opaque t :: %__MODULE__{}

  @spec create(Keyword.t()) :: {:ok, t} | {:error, term}
  def create(options \\ []) when is_list(options) do
    with {:ok, reference} <- NIF.make_vault(), do: {:ok, %__MODULE__{reference: reference}}
  end

  @spec sha256(t, binary) :: {:ok, binary} | {:error, term}
  def sha256(%__MODULE__{reference: vault_ref}, data) when is_binary(data),
    do: NIF.sha256(vault_ref, data)

  def generate_secret(%__MODULE__{reference: vault_ref}, attributes \\ []) do
    {type, attributes} = Keyword.pop(attributes, :type)
    attributes = if type, do: Keyword.put(attributes, :ty, type), else: attributes
    attributes = struct(SecretAttributes, attributes)

    with {:ok, secret_ref} <- NIF.generate_secret(vault_ref, attributes) do
      {:ok, %Secret{reference: secret_ref}}
    end
  end

  def generate_curve25519_keypair(vault) do
    with {:ok, private_key} <- generate_secret(vault, type: :curve25519_private),
         {:ok, public_key} <- get_public_key(vault, private_key) do
      {:ok, %{private: private_key, public: public_key}}
    end
  end

  def import_secret(%__MODULE__{reference: vault_ref}, secret, attributes \\ []) do
    {type, attributes} = Keyword.pop(attributes, :type)
    attributes = if type, do: Keyword.put(attributes, :ty, type), else: attributes
    attributes = struct(SecretAttributes, attributes)
    attributes = Map.put(attributes, :length, byte_size(secret))

    with {:ok, secret_ref} <- NIF.import_secret(vault_ref, secret, attributes) do
      {:ok, %Secret{reference: secret_ref}}
    end
  end

  def export_secret(%__MODULE__{reference: vault_ref}, %Secret{reference: secret_ref}),
    do: NIF.export_secret(vault_ref, secret_ref)

  def get_secret_attributes(%__MODULE__{reference: vault_ref}, %Secret{reference: secret_ref}) do
    with {:ok, attributes} <- NIF.get_secret_attributes(vault_ref, secret_ref) do
      {type, attributes} = attributes |> Map.from_struct() |> Map.to_list() |> Keyword.pop(:ty)
      if type, do: Keyword.put(attributes, :type, type), else: attributes
    end
  end

  def set_secret_type(%__MODULE__{reference: vault_ref}, %Secret{reference: secret_ref}, type),
    do: NIF.set_secret_type(vault_ref, secret_ref, type)

  @spec get_public_key(t, Secret.t()) :: {:ok, binary()} | {:error, term}
  def get_public_key(%__MODULE__{reference: vault_ref}, %Secret{reference: secret_ref}),
    do: NIF.get_public_key(vault_ref, secret_ref)

  def ecdh(%__MODULE__{reference: vault_ref}, %Secret{reference: private_key_ref}, peer_pubkey) do
    with {:ok, secret_ref} <- NIF.ecdh(vault_ref, private_key_ref, peer_pubkey) do
      {:ok, %Secret{reference: secret_ref}}
    end
  end

  def hkdf_sha256(
        %__MODULE__{reference: vault_ref},
        %Secret{reference: salt_ref},
        nil,
        num_outputs
      ) do
    with {:ok, outputs} <- NIF.hkdf_sha256(vault_ref, salt_ref, nil, num_outputs) do
      {:ok, Enum.map(outputs, fn x -> %Secret{reference: x} end)}
    end
  end

  def hkdf_sha256(
        %__MODULE__{reference: vault_ref},
        %Secret{reference: salt_ref},
        %Secret{reference: ikm_ref},
        num_outputs
      ) do
    with {:ok, outputs} <- NIF.hkdf_sha256(vault_ref, salt_ref, ikm_ref, num_outputs) do
      {:ok, Enum.map(outputs, fn x -> %Secret{reference: x} end)}
    end
  end

  def encrypt(
        %__MODULE__{reference: vault_ref},
        %Secret{reference: key_ref},
        nonce,
        aad,
        plaintext
      )
      when is_integer(nonce) do
    NIF.aead_aes_gcm_encrypt(vault_ref, key_ref, nonce, aad, plaintext)
  end

  def decrypt(
        %__MODULE__{reference: vault_ref},
        %Secret{reference: key_ref},
        nonce,
        aad,
        ciphertext_and_tag
      )
      when is_integer(nonce) do
    NIF.aead_aes_gcm_decrypt(vault_ref, key_ref, nonce, aad, ciphertext_and_tag)
  end

  defmodule NIF do
    @moduledoc false

    use Rustler, otp_app: :ockam, crate: :ockam_nif

    def make_vault, do: exit(:nif_not_loaded)
    def random(_vault), do: exit(:nif_not_loaded)
    def sha256(_vault, _data), do: exit(:nif_not_loaded)
    def generate_secret(_vault, _attrs), do: exit(:nif_not_loaded)
    def import_secret(_vault, _data, _attrs), do: exit(:nif_not_loaded)
    def export_secret(_vault, _secret), do: exit(:nif_not_loaded)
    def get_secret_attributes(_vault, _secret), do: exit(:nif_not_loaded)
    def set_secret_type(_vault, _secret, _secret_type), do: exit(:nif_not_loaded)
    def get_public_key(_vault, _secret), do: exit(:nif_not_loaded)
    def ecdh(_vault, _private_key, _peer_pubkey), do: exit(:nif_not_loaded)
    def hkdf_sha256(_vault, _salt, _ikm, _num_derived_outputs), do: exit(:nif_not_loaded)

    def aead_aes_gcm_encrypt(_vault, _key, _nonce, _additional_data, _plaintext),
      do: exit(:nif_not_loaded)

    def aead_aes_gcm_decrypt(_vault, _key, _nonce, _additional_data, _ciphertext_and_tag),
      do: exit(:nif_not_loaded)
  end
end

File: implementations/rust/ockam/src/vault.rs
use crate::error::Result as OckamResult;

cfg_if!{
    if #[cfg(feature = "heapless")] {
        use crate::heapless::consts::*;
        pub type SecretKey = heapless::Vec<u8, U32>;
        pub type PublicKey = heapless::Vec<u8, U65>;
        pub type SmallBuffer<T> = heapless::Vec<T, U4>;
        pub type Buffer<T> = heapless::Vec<T, U512>;
    }
    else {
        pub type SecretKey = Vec<u8>;
        pub type PublicKey = Vec<u8>;
        pub type SmallBuffer<T> = Vec<T>;
        pub type Buffer<T> = Vec<T>;
    }
}

/// Represents a secret handle or context that is
/// stored in a vault
pub trait Secret: std::fmt::Debug + Send + downcast::Any + Sync + 'static {}
downcast!(dyn Secret);

/// Persistence allowed by Secrets
#[derive(Copy, Clone, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
pub enum SecretPersistence {
    Ephemeral,
    Persistent
}

/// The types of secret keys that the vault supports
#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
pub enum SecretType {
    /// Raw buffer of bytes
    Buffer,
    /// AES key
    Aes,
    /// x25519 secret key
    Curve25519,
    /// NIST P-256 (secp256r1, prime256v1) secret key
    P256,
}

/// Attributes for a specific vault secret
#[derive(Copy, Clone, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
pub struct SecretAttributes {
    /// The type of key
    pub stype: SecretType,
    /// How the key is persisted
    pub persistence: SecretPersistence,
    /// The purpose of the secret key
    pub length: u32,
}

pub trait SecretVault {
    /// Create a new secret key
    fn secret_generate(
        &mut self,
        attributes: SecretAttributes,
    ) -> OckamResult<Box<dyn Secret>>;
    /// Import a secret key into the vault
    fn secret_import(
        &mut self,
        secret: &SecretKey,
        attributes: SecretAttributes,
    ) -> OckamResult<Box<dyn Secret>>;
    /// Return the associated public key given the secret key
    fn secret_public_key_get(
        &mut self,
        context: &Box<dyn Secret>,
    ) -> OckamResult<PublicKey>;
}

pub trait SigningVault {
    /// Generate a signature
    fn sign(
        &mut self,
        secret_key: &Box<dyn Secret>,
        data: &[u8],
    ) -> OckamResult<[u8; 64]>;
    /// Verify a signature
    fn verify(
        &mut self,
        signature: [u8; 64],
        public_key: PublicKey,
        data: &[u8],
    ) -> OckamResult<()>;
}

pub trait SymmetricVault {
    /// Compute the SHA-256 digest given input `data`
    fn sha256(&self, data: &[u8]) -> OckamResult<[u8; 32]>;
    /// Compute Elliptic-Curve Diffie-Hellman using this secret key
    /// and the specified uncompressed public key
    fn ec_diffie_hellman(
        &mut self,
        context: &Box<dyn Secret>,
        peer_public_key: PublicKey,
    ) -> OckamResult<Box<dyn Secret>>;
    /// Compute the HKDF-SHA256 using the specified salt and input key material
    /// and return the output key material of the specified length
    fn hkdf_sha256(
        &mut self,
        salt: &Box<dyn Secret>,
        info: &[u8],
        ikm: Option<&Box<dyn Secret>>,
        output_attributes: Vec<SecretAttributes>,
    ) -> OckamResult<SmallBuffer<Box<dyn Secret>>>;
    /// Encrypt a payload using AES-GCM
    fn aead_aes_gcm_encrypt(
        &mut self,
        context: &Box<dyn Secret>,
        plaintext: &[u8],
        nonce: &[u8],
        aad: &[u8],
    ) -> OckamResult<Buffer<u8>>;
    /// Decrypt a payload using AES-GCM
    fn aead_aes_gcm_decrypt(
        &mut self,
        context: &Box<dyn Secret>,
        cipher_text: &[u8],
        nonce: &[u8],
        aad: &[u8],
    ) -> OckamResult<Buffer<u8>>;
}